# Tarea 1 Bases de Datos 2023-1 Grupo 11


## Integrantes:
* Bastían Wohlwend - 202173006-0
* Scarleth Bazaes - 202173015-k

## Requisitos para Correr el Notebook

Para poder utilizar este Notebook se debe tener instalado:
- **Python 3**
- **Jupyter**
- La librería **ipython-sql** (Poder ejecutar sql en Jupyter Notebook)
- PostgreSQL y su conector desde Python(**psycopg2**)

**IMPORTANTE:** Este Jupyter Notebook se conectará a su servicio de Base de Datos PostgreSQL de manera **local**, por lo que **no funcionará correctamente en el entorno Google Collab.**

Para instalar las dependencias, ejecute (**recuerde actualizar pip3**):

In [1]:
! pip3 install ipython-sql
! pip3 install psycopg2

#Use to generate names
! pip3 install names

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 23.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importar dependencias

In [2]:
from datetime import datetime
import random as r
import names

#Librerías para PostgreSQL
import psycopg2

Para conectarse a la base de datos, completen los datos relevantes en la variable DATABASE_URL.

* Cambie `user` por el usuario que utilizó en Postgres para acceder a la Base de Datos. Por defecto `postgres`.
* Cambie `password` por la contraseña de dicho usuario. **Recuerde borrarla antes de entregar su tarea.**
* Cambie `port` por el puerto en que se ejecuta su servicio de Base de Datos. Por defecto `5432`.
* Cambie `database` por el nombre de la BD (Schema) que creó en Postgres.

Ejemplo: `postgresql://postgres:pass_segura@localhost:5432/tarea-bd`

In [3]:
%reload_ext sql
%env DATABASE_URL=postgresql://postgres:Leprechaun10@localhost:5432/tarea1_bd

env: DATABASE_URL=postgresql://postgres:Leprechaun10@localhost:5432/tarea1_bd


Para conectarse de forma declarativa a la Base de Datos, utilice el siguiente código:

In [4]:
#Recuerde cambiar los valores de USER, DBNAME y PASSWORD por los mismos de la celda superior.
connection = psycopg2.connect(
   host="localhost",
   user="postgres",
   dbname='tarea1_bd',
   password="Leprechaun10"
)

print(connection)
cursor = connection.cursor()

<connection object at 0x000001CEE9C597A0; dsn: 'user=postgres password=xxx dbname=tarea1_bd host=localhost', closed: 0>


## Crear tablas (36 pts.)

Aquí inserta el código SQL que creara las tablas de la base de datos:

### Tabla "facciones" (6 pts)

In [5]:
%%sql

CREATE TABLE facciones (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    PRIMARY KEY (id)
);

(psycopg2.errors.DuplicateTable) la relación «facciones» ya existe

[SQL: CREATE TABLE facciones (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    PRIMARY KEY (id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "trabajos" (6 pts)

In [6]:
%%sql

CREATE TABLE trabajos (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL,
    PRIMARY KEY (id)
);


 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.DuplicateTable) la relación «trabajos» ya existe

[SQL: CREATE TABLE trabajos (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    descripcion VARCHAR(45) NOT NULL,
    sueldo INTEGER NOT NULL,
    PRIMARY KEY (id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "personas" (6 pts)

In [7]:
%%sql

CREATE TABLE personas (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
    infectado BOOL NOT NULL,
    conyugue INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (conyugue) REFERENCES personas(id)
);

 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.DuplicateTable) la relación «personas» ya existe

[SQL: CREATE TABLE personas (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    nombre VARCHAR(45) NOT NULL,
    apellidos VARCHAR(45) NOT NULL,
    fecha_nacimiento DATE NOT NULL,
    infectado BOOL NOT NULL,
    conyugue INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (conyugue) REFERENCES personas(id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "cuentas_bancarias" (6 pts)

In [8]:
%%sql

CREATE TABLE cuentas_bancarias (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    tipo_de_cuenta VARCHAR(45) NOT NULL,
    banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (id_persona) REFERENCES personas(id)
);

 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.DuplicateTable) la relación «cuentas_bancarias» ya existe

[SQL: CREATE TABLE cuentas_bancarias (
    id INTEGER GENERATED BY DEFAULT AS IDENTITY,
    tipo_de_cuenta VARCHAR(45) NOT NULL,
    banco VARCHAR(45) NOT NULL,
    saldo INTEGER NOT NULL,
    id_persona INTEGER,
    PRIMARY KEY (id),
    FOREIGN KEY (id_persona) REFERENCES personas(id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "persona_pertenece_faccion" (6 pts)

In [9]:
%%sql

CREATE TABLE persona_pertenece_faccion (
    id_persona INTEGER GENERATED BY DEFAULT AS IDENTITY,
    id_faccion INTEGER GENERATED BY DEFAULT AS IDENTITY,
    PRIMARY KEY (id_persona,id_faccion),
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_faccion) REFERENCES facciones(id)
);


 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.DuplicateTable) la relación «persona_pertenece_faccion» ya existe

[SQL: CREATE TABLE persona_pertenece_faccion (
    id_persona INTEGER GENERATED BY DEFAULT AS IDENTITY,
    id_faccion INTEGER GENERATED BY DEFAULT AS IDENTITY,
    PRIMARY KEY (id_persona,id_faccion),
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_faccion) REFERENCES facciones(id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


### Tabla "persona_tiene_trabajo" (6 pts)

In [10]:
%%sql

CREATE TABLE persona_tiene_trabajo (
    id_persona INTEGER GENERATED BY DEFAULT AS IDENTITY,
    id_trabajo INTEGER GENERATED BY DEFAULT AS IDENTITY,
    estado BOOL NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    PRIMARY KEY (id_persona,id_trabajo),
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_trabajo) REFERENCES trabajos(id)
);

 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.DuplicateTable) la relación «persona_tiene_trabajo» ya existe

[SQL: CREATE TABLE persona_tiene_trabajo (
    id_persona INTEGER GENERATED BY DEFAULT AS IDENTITY,
    id_trabajo INTEGER GENERATED BY DEFAULT AS IDENTITY,
    estado BOOL NOT NULL,
    ultima_vez_realizado TIMESTAMP NOT NULL,
    PRIMARY KEY (id_persona,id_trabajo),
    FOREIGN KEY (id_persona) REFERENCES personas(id),
    FOREIGN KEY (id_trabajo) REFERENCES trabajos(id)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


## Código Generador de Datos

Ejecute este código para cargar datos a las tablas creadas.

PD: Si el codigo tira error por "violar la unicidad de las PK" es por una coincidencia de numeros random, solamente volver a ejecutar el codigo

In [11]:
#Recuerde haber ejecutado todas las celdas anteriores!

# Generador de timestamp aleatorio.
def generate_timestamp_nacimiento():
    year = r.randint(1970,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

def generate_timestamp_trabajo():
    year = r.randint(2019,2022)
    month = r.randint(1,12)
    day = r.randint(1,31)
    
    hour = r.randint(0,23)
    minute = r.randint(0,59)
    second = r.randint(0,59)
    
    if month in [4, 6, 9, 11] and day > 30:
        day = 30
        
    if month == 2 and day > 28:
        day = 28
        
    return datetime(year, month, day, hour, minute, second)

print("Borrando Datos Antiguos...")

# Limpia la base de datos.
connection.commit() # En caso de algun error
query = 'TRUNCATE TABLE "facciones", "trabajos", "personas", "cuentas_bancarias", "persona_pertenece_faccion", "persona_tiene_trabajo" RESTART IDENTITY'
cursor.execute(query)
connection.commit()

print("Generando Datos...")

facciones = [("FEDRA", "La Agencia Federal de Respuesta a Desastres"), ("Luciérnagas", "Grupo de milicias revolucionario"), ("Carroñeros", "Sin identificación"), ("Pythonia", "Grupo defensor de la libertad"), ("MaCdonia", "Los exiliados de la carceles"), ("JavaScriptia", "Grupo defensor de la paz")]
trabajos = [("Limpieza", "Encargado de limpiar las calles"),
("Limpiador de cloacas","Encargado de limpiar las cloacas"),
("Militar","Encargado de proteger la ciudad"),
("Explorador","Encargado de explorar la ciudad"),
("Médico","encargado de curar a los enfermos"),
("Profesor","Encargado de enseñar a los niños"),
("Cremador","Encargado de quemar los cadáveres"),
("Científico","Encargado de investigar"),
("Cocinero","Encargado de cocinar"),
("Carpintero","Encargado de construir"),
("Cerrajero","Encargado de reparar las cerraduras"),
("Cartero","Encargado de repartir los correos"),
("Cajero","Encargado de cobrar los impuestos")]
tipos_de_cuenta_bancaria = ["Cuenta Vista", "Cuenta Ahorro", "Cuenta Corriente", "Cuenta RUT", "Cuenta de Inversion", "Cuenta PrePago", "Cuenta de Credito"]
bancos = ["FedraBank S.A", "Banco Pedrito Pascal SpA", "Banco Pythonia S.A", "Banco de la Nación S.A", "Banco Santander SpA", "Banco BCI SpA", "Banco Estado SpA", "Banco Consorcio S.A", "Banco BBVA", "Banco Scotiabank"]

# Variables.
n_facciones = len(facciones)
n_trabajos = len(trabajos)
n_personas = 1000
n_cuentas_bancarias = 120
n_persona_pertenece_faccion = 300
n_persona_tiene_trabajo = 300
n_casados = 100

# Genera personas.
personas = []
for _ in range(n_personas):
    nombre = names.get_first_name()
    apellidos = names.get_last_name()+" "+names.get_last_name()
    fecha_nacimiento = generate_timestamp_nacimiento()
    infectado = r.choice([True, False])
    
    personas.append((nombre, apellidos, fecha_nacimiento, infectado, None))


# Genera cuentas bancarias.
cuentas_bancarias = []
for _ in range(n_cuentas_bancarias):
    tipo_de_cuenta = r.choice(tipos_de_cuenta_bancaria)
    banco = r.choice(bancos)
    saldo = r.randint(0, 10000)
    id_persona = r.choice(range(1, n_personas+1))
    
    cuentas_bancarias.append((tipo_de_cuenta, banco, saldo, id_persona))
    
    
# Generate persona_pertenece_faccion.
persona_pertenece_faccion = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_facciones+1))) for _ in range(1, n_persona_pertenece_faccion+1)])

# Generate persona_tiene_trabajo.
persona_tiene_trabajo_prev = set([(r.choice(range(1, n_personas+1)), r.choice(range(1, n_trabajos+1))) for _ in range(1, n_persona_tiene_trabajo+1)])
persona_tiene_trabajo = [(id_persona, id_trabajo, r.choice([True, False]), generate_timestamp_trabajo()) for id_persona, id_trabajo in persona_tiene_trabajo_prev]
    
# SQL Statements.
inserts = [
    'INSERT INTO "facciones" ("nombre", "descripcion") VALUES (%s, %s);',
    'INSERT INTO "trabajos" ("nombre", "descripcion", "sueldo") VALUES (%s, %s, %s);',
    'INSERT INTO "personas" ("nombre", "apellidos", "fecha_nacimiento", "infectado", "conyugue") VALUES (%s, %s, %s, %s, %s);',
    'INSERT INTO "cuentas_bancarias" ("tipo_de_cuenta", "banco", "saldo", "id_persona") VALUES (%s, %s, %s, %s);',
    'INSERT INTO "persona_pertenece_faccion" ("id_persona", "id_faccion") VALUES (%s, %s);',
    'INSERT INTO "persona_tiene_trabajo" ("id_persona", "id_trabajo", "estado", "ultima_vez_realizado") VALUES (%s, %s, %s, %s);'
]

for faccion in facciones:
    cursor.execute(inserts[0], faccion)

for trabajo in trabajos:
    t = (trabajo[0], trabajo[1], r.randint(100, 5000))
    cursor.execute(inserts[1], t)

# Crear una lista con personas casadas y otra con personas solteras. donde se cambie la ide de la persona que se casó.

for persona in personas:
    cursor.execute(inserts[2], persona)
    
for cb in cuentas_bancarias:
    cursor.execute(inserts[3], cb)

for ppf in persona_pertenece_faccion:
    cursor.execute(inserts[4], ppf)
    
for ptt in persona_tiene_trabajo:
    cursor.execute(inserts[5], ptt)
    
connection.commit()


# Generar Matrimonios
n_repetidos = []

for i in range(n_casados):
    random = r.randint(1, n_personas)
    if random in n_repetidos:
        continue
    else:
        random2 = r.randint(1, n_personas)
        if random2 in n_repetidos:
            continue
        else:
            n_repetidos.append(random)
            n_repetidos.append(random2)
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random2, random))
            cursor.execute("UPDATE personas SET conyugue = %s WHERE id = %s", (random, random2))
            connection.commit()

print("Todo Listo!")

Borrando Datos Antiguos...
Generando Datos...
Todo Listo!


## Consultas (64 pts.)

Escriba aquí las consultas pedidas. Recuerde agregar una descripción _breve_ de cómo funciona la solución. **Escriba las consultas de forma ordenada.**

**NOTA:** Cuando en una consulta diga "X cosa" significa que ustedes deben de elegir ese X por lo que para cualquier otro X debe de funcionar de igual forma.

### Consulta 1 (6 pts)
**ID, nombre y primer apellido de los habitantes de FEDRA**

Ejemplo:

|ID |  Nombre      |  Apellido  |
|---|-----------|-----------|
| 12 |Joel | Miller |
|21 | Ellie | Williams |
| ... | ... | ... |

**Solución:** _Realizamos un inner join entre las tablas "personas", "persona_pertenece_faccion" y "facciones", del cual seleccionamos el id, nombre, y apellido, recortando apellidos hasta donde encuentre un espacio en blanco. Finalmente ponemos como condicion que estas personas pertenezcan a FEDRA._


In [25]:
%%sql

SELECT 
	personas.id AS "ID", 
	personas.nombre AS "Nombre", 
	SUBSTRING(personas.apellidos,0,strpos(personas.apellidos,' ')) "Apellido"
FROM 
	personas 
	INNER JOIN persona_pertenece_faccion
	ON personas.id = persona_pertenece_faccion.id_persona
	INNER JOIN facciones
	ON facciones.id = persona_pertenece_faccion.id_faccion
WHERE 
	facciones.nombre = 'FEDRA'
ORDER BY 
	personas.id;


 * postgresql://postgres:***@localhost:5432/tarea1_bd
51 rows affected.


[(3, 'Samuel', 'Honor'),
 (11, 'Alica', 'Tidwell'),
 (40, 'James', 'Clyburn'),
 (65, 'Charles', 'Treat'),
 (82, 'Charline', 'Carter'),
 (149, 'Yasmin', 'Thompson'),
 (158, 'Judith', 'Weslowski'),
 (162, 'Sandra', 'Brown'),
 (178, 'Veronica', 'Young'),
 (195, 'Clark', 'Campos'),
 (204, 'Marilyn', 'Baldwin'),
 (207, 'Jocelyn', 'Sutherland'),
 (216, 'Mary', 'Mullins'),
 (221, 'Lucille', 'Ham'),
 (229, 'Veronica', 'Spellman'),
 (269, 'Shannon', 'Whitlock'),
 (289, 'Tara', 'Smith'),
 (311, 'Jon', 'Raman'),
 (338, 'Terrie', 'Mcgehee'),
 (379, 'Brittany', 'Laman'),
 (388, 'John', 'Perez'),
 (416, 'Ronald', 'Felmet'),
 (430, 'Richard', 'Fletcher'),
 (437, 'Clarisa', 'Inverso'),
 (477, 'Sheila', 'Lopez'),
 (491, 'Bobbie', 'Lotts'),
 (508, 'Shawnta', 'Zhong'),
 (539, 'Larry', 'Miller'),
 (544, 'Louis', 'Gittens'),
 (602, 'Martha', 'Anderson'),
 (616, 'Jenna', 'Wilson'),
 (637, 'Jocelyn', 'Hildebrand'),
 (639, 'Anthony', 'Wilson'),
 (649, 'Bonnie', 'Beard'),
 (660, 'Jacqueline', 'Alexander'),
 (667, 'Clarence', 'Miller'),
 (676, 'Danny', 'Tipka'),
 (693, 'Anthony', 'Anaya'),
 (704, 'Chadwick', 'Davis'),
 (731, 'Eugene', 'Garrity'),
 (741, 'Dorothy', 'Lemons'),
 (754, 'Teresa', 'Boas'),
 (785, 'Jarrod', 'Gallardo'),
 (838, 'Mary', 'Culbreth'),
 (857, 'Jerry', 'Flores'),
 (912, 'John', 'Medrano'),
 (922, 'Kristin', 'Hicks'),
 (958, 'Gigi', 'Rodriquez'),
 (970, 'Claudia', 'Jones'),
 (984, 'Carolyn', 'Henry'),
 (997, 'Curtis', 'Mermis')]

### Consulta 2 (6 pts)
**Cantidad de personas nacidas después del inicio de la pandemia.**(inicio de la pandemia 01-12-2019)

Ejemplo:


| Cantidad     | 
|--------------|
| 292 | 

**Solución:** _Contamos la cantidad de id's que existen de personas cuya fecha de nacimiento sea mayor a la correspondiente al inicio de la pandemia_

In [26]:
%%sql

SELECT COUNT(id) AS "Cantidad"
FROM personas
WHERE fecha_nacimiento > '2019-12-01'

 * postgresql://postgres:***@localhost:5432/tarea1_bd
1 rows affected.


[(52,)]

### Consulta 3 (6 pts)
**ID, nombre, apellidos y saldo total de todas las personas**( Limitar a 20 filas para visualizar)

Ejemplo:

| ID    | Nombre | Apellido | Dinero | 
|----------------|--------|-|-|
| 13 | Tommy | Miller | 12501 |
| ... | ... | ... | ... |

**Solución:** _Realizamos un inner join entre personas y cuentas_bancarias, seleccionamos la informacion pedida, en el caso de "Apellido" realizamos el corte necesario al string, y limitamos la cantidad de filas a 20_

In [29]:
%%sql

SELECT	
	personas.id AS "ID",
	personas.nombre AS "Nombre", 
	SUBSTRING(personas.apellidos,0,strpos(personas.apellidos,' ')) "Apellido",
	cuentas_bancarias.saldo AS "Dinero"
FROM 
	personas INNER JOIN cuentas_bancarias
	ON personas.id = cuentas_bancarias.id_persona
ORDER BY
	saldo DESC
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
20 rows affected.


[(954, 'David', 'Garza', 9993),
 (925, 'Leonard', 'Tribble', 9966),
 (630, 'David', 'Hayes', 9936),
 (244, 'Gary', 'Warnock', 9929),
 (710, 'Alison', 'Tone', 9912),
 (382, 'Helen', 'Pope', 9901),
 (375, 'Karen', 'Duwe', 9868),
 (716, 'Jay', 'Jones', 9843),
 (987, 'Maria', 'Kirkpatrick', 9810),
 (956, 'Vickie', 'Mompoint', 9574),
 (411, 'James', 'Cunningham', 9536),
 (764, 'Ida', 'Teeter', 9488),
 (785, 'Jarrod', 'Gallardo', 9481),
 (213, 'Denise', 'Hall', 9267),
 (266, 'Shirley', 'Upshaw', 9249),
 (462, 'John', 'Kohl', 9208),
 (486, 'Ryan', 'Mahon', 8907),
 (608, 'Jacob', 'Bradley', 8894),
 (34, 'Laura', 'Lucero', 8848),
 (554, 'Jonathan', 'Savala', 8842)]

### Consulta 4 (6 pts)
**Nombre de la Facción y la cantidad de trabajadores activos hasta fecha X.**

Ejemplo:

| Facción     | Cantidad de trabajadores |
|----------------|--------|
| FEDRA | 1273 |
| Luciérnagas | 121 |
| Carroñeros  | 72 |
| ...  | ... |

**Solución:** _Realizamos un inner join entre personas, persona_pertenece_faccion, facciones y persona_tiene_trabajo. Pedimos la faccion y la cantidad de trabajadores, la cual la calculamos haciendo un count a persona_pertenece_faccion.id_persona. Ponemos como condición que la persona haya nacido antes de la fecha pedida y que esta esté actualmente trabajando._

In [31]:
%%sql


SELECT 
	facciones.nombre AS "Facción", 
	COUNT(persona_pertenece_faccion.id_persona) AS "Cantidad de trabajadores"
FROM
	personas INNER JOIN persona_pertenece_faccion
	ON personas.id = persona_pertenece_faccion.id_persona
	INNER JOIN facciones
	ON persona_pertenece_faccion.id_faccion = facciones.id
	INNER JOIN persona_tiene_trabajo
	ON personas.id = persona_tiene_trabajo.id_persona
WHERE personas.fecha_nacimiento < '2015-10-06' AND persona_tiene_trabajo.estado = TRUE
GROUP BY facciones.nombre;


 * postgresql://postgres:***@localhost:5432/tarea1_bd
6 rows affected.


[('JavaScriptia', 7),
 ('Pythonia', 3),
 ('MaCdonia', 6),
 ('FEDRA', 7),
 ('Luciérnagas', 8),
 ('Carroñeros', 5)]

### Consulta 5 (6 pts)
**Nombre de la Facción y su cantidad de personas infectadas.**

Ejemplo:

|  Facción  | Personas infectadas | 
|--------|-------|
| FEDRA | 251 |
| Luciérnagas | 17 |
| Carroñeros  | 12 |
| ...  | ... |

**Solución:** _Primeramente explicaremos las inner join de las consultas:

Generamos una tabla seleccionando los atributos id_persona e id_faccion desde un INNER JOIN entre las tablas "personas" y "persona_pertenece_facción", considerando sólo aquellas personas que estén infectadas a la cual llamamos sub_union

Luego, desde un INNER JOIN entre "sub_union" y "facciones" generamos otra tabla seleccionando el nombre (de la facción), la cantidad de personas infectadas (COUNT id_persona), agrupamos la tabla anterior por el nombre de la faccion y la ordenamos de manera descendiente según cantidad de personas infectadas._

In [33]:
%%sql

SELECT
    nombre AS "Facción", COUNT(id_persona) AS "Personas infectadas"
FROM(
    SELECT 
        id_persona, id_faccion
    FROM
        personas INNER JOIN persona_pertenece_faccion on id = id_persona
    WHERE
        infectado = True
    )
    AS sub_union INNER JOIN facciones on id_faccion = id

GROUP BY
    nombre 

ORDER BY
    COUNT(id_persona) DESC;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
6 rows affected.


[('FEDRA', 33),
 ('Carroñeros', 26),
 ('JavaScriptia', 23),
 ('MaCdonia', 22),
 ('Pythonia', 21),
 ('Luciérnagas', 16)]

### Consulta 6 (6 pts)
**Nombre de la Facción y su trabajo más realizado.**

Ejemplo:

|  Facción  | Trabajo | 
|--------|-------|
| FEDRA | Limpiador de cloacas |
| Luciérnagas | Médico |
| Carroñeros  | Explorador |
| ...  | ... |

**Solución:** _Seleccionamos trabajo, cantidad_trabajadores y faccion desde un inner join de personas, persona_tiene_trabajo, trabajos, persona_pertenece_faccion y facciones (gigante), esta tabla la llamamos nueva_tabla y nos muestra la cantidad de trabajadores de cada oficio por facción. Luego de esta tabla seleccionamos trabajo, cantidad_trabajadores y faccion donde la cantidad de trabajadores sea maxima, para esto utilizamos un ANY dentro del cual se encuentra una tabla que contiene las facciones y su respectiva mayor cantidad de trabajadores. Finalemente (T-T) seleccionamos faccion y trabajo, como algunos trabajos empatan en cantidad maxima de trabajadores para cada faccion, escogemos uno utilizando la funcion MAX en trabajo._

In [17]:
%%sql

SELECT  tablita.faccion AS "Facción", 
		MAX(tablita.trabajo) AS "Trabajo"
FROM (

    SELECT
        nueva_tabla.faccion AS faccion,
        nueva_tabla.cantidad_trabajadores AS cantidad_trabajadores,
        nueva_tabla.trabajo AS trabajo
    FROM(
        SELECT
            trabajos.nombre AS trabajo,
            COUNT(persona_tiene_trabajo.id_persona) AS cantidad_trabajadores,
            facciones.nombre AS faccion
        FROM
            personas INNER JOIN persona_tiene_trabajo
            ON personas.id = persona_tiene_trabajo.id_persona
            INNER JOIN trabajos
            ON persona_tiene_trabajo.id_trabajo = trabajos.id
            INNER JOIN persona_pertenece_faccion
            ON personas.id= persona_pertenece_faccion.id_persona
            INNER JOIN facciones
            ON persona_pertenece_faccion.id_faccion = facciones.id
        GROUP BY facciones.nombre, trabajos.nombre
        ) AS nueva_tabla
    WHERE (nueva_tabla.faccion,nueva_tabla.cantidad_trabajadores) = ANY(
        SELECT
            nueva_tabla1.faccion AS "Faccion",
            MAX(nueva_tabla1.cantidad_trabajadores) AS "Cantidad Trabajadores"
        FROM(
            SELECT
                trabajos.nombre AS trabajo,
                COUNT(persona_tiene_trabajo.id_persona) AS cantidad_trabajadores,
                facciones.nombre AS faccion
            FROM
                personas INNER JOIN persona_tiene_trabajo
                ON personas.id = persona_tiene_trabajo.id_persona
                INNER JOIN trabajos
                ON persona_tiene_trabajo.id_trabajo = trabajos.id
                INNER JOIN persona_pertenece_faccion
                ON personas.id= persona_pertenece_faccion.id_persona
                INNER JOIN facciones
                ON persona_pertenece_faccion.id_faccion = facciones.id
            GROUP BY facciones.nombre, trabajos.nombre
            ) AS nueva_tabla1
            GROUP BY nueva_tabla1.faccion
            )) AS tablita
GROUP BY tablita.faccion;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
6 rows affected.


[('Carroñeros', 'Cartero'),
 ('FEDRA', 'Cerrajero'),
 ('JavaScriptia', 'Explorador'),
 ('Luciérnagas', 'Cerrajero'),
 ('MaCdonia', 'Militar'),
 ('Pythonia', 'Profesor')]

### Consulta 7 (8 pts)
**Nombre de la Facción y su Saldo total** (Es decir, la suma total de los saldos de cada persona perteneciente a la facción.)

Ejemplo:

|Facción | Saldo |
|--------|-|
| FEDRA | 523000 |
| Luciérnagas | 130023 |
| Carroñeros  | 17432 |
| ...  | ...|

**Solución:** _Primeramente explicaremos los inner joins interiores:

Generamos una tabla a partir de un INNER JOIN entre las tablas "cuentas_bancarias" y "persona_pertenece_faccion" conisderando los id de personas, Agrupamos considerando el id_faccion, y seleccinoamos id_faccion y la suma de los saldos por cada faccion, la tabla anterior la llamamos sub_union.

Realizamos un INNER JOIN entre las tablas "sub_union" y "facciones" considerando los id de las facciones, seleccionamos el nombre (de la faccion) y la suma de los saldo por faccion ("Saldo")
además, para mayor orden ordenamos los saldos de manera descendiente._

In [34]:
%%sql

SELECT
    nombre AS "Facción", "Saldo"
FROM(
    SELECT
        id_faccion, SUM(saldo) AS "Saldo"
    FROM
        cuentas_bancarias INNER JOIN persona_pertenece_faccion ON cuentas_bancarias.id_persona = persona_pertenece_faccion.id_persona
    GROUP BY
        id_faccion
    ) AS sub_union

    INNER JOIN facciones on id_faccion = id

ORDER BY
    "Saldo" DESC;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
6 rows affected.


[('JavaScriptia', 38597),
 ('MaCdonia', 35904),
 ('Luciérnagas', 28672),
 ('Pythonia', 27998),
 ('Carroñeros', 22629),
 ('FEDRA', 19289)]

### Consulta 8 (6 pts)
**Nombre del Banco, cantidad de usuarios pertenecientes al banco y saldo total del banco.**

###### Es decir, todos los bancos que ocupan los habitantes de cada faccion

Ejemplo:

|Banco | Cantidad|Saldo|
|----|---|---|
|FedraBank S.A |321|20000|
|Banco Pedrito Pascal SpA  |21|200| 
| Banco De Chile|213|56000|
| ... | ... |

**Solución:** _Realizamos un inner join entre personas y cuentas_bancarias, del cual seleccionamos el banco, la cantidad de cuentas bancarias utilizando una función count en cuentas_bancarias.id_persona, y el saldo total de este utilizando una función SUM en cuentas_bancarias.saldo, todo esto lo agrupamos por banco y lo ordenamos por saldo._

In [53]:
%%sql

SELECT
	cuentas_bancarias.banco AS "Banco",
	COUNT(cuentas_bancarias.id_persona) AS "Cantidad",
	SUM(cuentas_bancarias.saldo) AS "Saldo"
FROM
	personas INNER JOIN cuentas_bancarias
	ON personas.id = cuentas_bancarias.id_persona
GROUP BY
	cuentas_bancarias.banco
ORDER BY
	"Saldo" DESC;



 * postgresql://postgres:***@localhost:5432/tarea1_bd
10 rows affected.


[('Banco Pythonia S.A', 19, 115847),
 ('Banco Pedrito Pascal SpA', 13, 77847),
 ('Banco Estado SpA', 13, 73622),
 ('Banco BBVA', 14, 70368),
 ('Banco Santander SpA', 11, 69024),
 ('FedraBank S.A', 12, 67871),
 ('Banco de la Nación S.A', 14, 64241),
 ('Banco BCI SpA', 10, 45583),
 ('Banco Consorcio S.A', 8, 44449),
 ('Banco Scotiabank', 6, 28157)]

### Consulta 9 (8 pts)
**Nombre de la Facción y el Banco más utilizado por sus habitantes**(Si hay empate, que se muestre solo uno.)

Ejemplo:
   
|Facción | Banco |
| - | - |
| FEDRA | FedraBank S.A |
| Luciérnagas | Banco Pedrito Pascal SpA |
| ... | ... |

**Solución:** _Realizamos un inner join entre personas, cuentas_bancarias, persona_pertenece_faccion y faccion, del cual seleccionamos faccion, cantidad de usuarios realizando un count a cuentas_bancarias.id_persona y banco. A esto lo llamamos nueva_tabla, y seleccionamos de ella la mayor cantidad de usuarios por banco y faccion utilizando un where donde faccion y cantidad de usuarios sean igual a un any dentro del cual se encuentra una tabla que contiene las facciones y su maxima cantidad de usuarios. Finalmente, similar a la consulta 6, pueden existir empates en la cantidad de usuarios maxima por faccion, por lo que seleccionamos una sola utilizando la funcion max en el nombre de los bancos._

In [52]:
%%sql

SELECT  tablita.faccion AS "Facción", 
        MAX(tablita.banco) AS "Banco"
FROM (

    SELECT
        nueva_tabla.faccion AS faccion,
        nueva_tabla.cantidad_usuarios AS cantidad_usuarios,
        nueva_tabla.banco AS banco
    FROM(
        SELECT
            facciones.nombre AS faccion,
            COUNT(cuentas_bancarias.id_persona) AS cantidad_usuarios,
            cuentas_bancarias.banco AS banco
        FROM
            personas INNER JOIN cuentas_bancarias
            ON personas.id = cuentas_bancarias.id_persona
            INNER JOIN persona_pertenece_faccion
            ON personas.id= persona_pertenece_faccion.id_persona
            INNER JOIN facciones
            ON persona_pertenece_faccion.id_faccion = facciones.id
        GROUP BY facciones.nombre, cuentas_bancarias.banco
        ) AS nueva_tabla
    WHERE (nueva_tabla.faccion,nueva_tabla.cantidad_usuarios) = ANY(
        SELECT
            nueva_tabla1.faccion AS "Faccion",
            MAX(nueva_tabla1.cantidad_usuarios) AS "Cantidad Usuarios"
        FROM(
            SELECT
                facciones.nombre AS faccion,
                COUNT(cuentas_bancarias.id_persona) AS cantidad_usuarios,
                cuentas_bancarias.banco AS banco
            FROM
                personas INNER JOIN cuentas_bancarias
                ON personas.id = cuentas_bancarias.id_persona
                INNER JOIN persona_pertenece_faccion
                ON personas.id= persona_pertenece_faccion.id_persona
                INNER JOIN facciones
                ON persona_pertenece_faccion.id_faccion = facciones.id
            GROUP BY facciones.nombre, cuentas_bancarias.banco
            ) AS nueva_tabla1
            GROUP BY nueva_tabla1.faccion
            )) AS tablita
GROUP BY tablita.faccion;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
6 rows affected.


[('Carroñeros', 'Banco Estado SpA'),
 ('FEDRA', 'FedraBank S.A'),
 ('JavaScriptia', 'Banco Santander SpA'),
 ('Luciérnagas', 'Banco Estado SpA'),
 ('MaCdonia', 'Banco Pythonia S.A'),
 ('Pythonia', 'Banco BBVA')]

### Consulta 10 (6 pts)
**ID, nombre y segundo apellido de las personas traidoras**(Es decir, si la persona pertenece a dos o más facciones.)

Ejemplo:
   
|ID | Nombre | Apellido | 
|-|-------|-|
| 1 | Pedro | Pascal |
| 2 | Bella | Ramsey |
| ... | ... | ... | 

**Solución:** _Realizamos un inner join entre personas y persona_pertenece_faccion, de este selecciamos ID, nombre y apellido (realizando el respectivo corte para que este corresponda al segundo) con la condicion de que la cantidad de relaciones persona_pertenece_faccion.id_faccion de cada persona sea mayor a 1. Luego agrupamos por id, nombre y apellido._

In [21]:
%%sql

SELECT
	personas.id AS "ID",
	personas.nombre AS "Nombre",
	SUBSTRING(personas.apellidos,strpos(personas.apellidos,' ')+1,LENGTH(personas.apellidos)) AS "Apellido"
FROM
	personas INNER JOIN persona_pertenece_faccion
	ON personas.id = persona_pertenece_faccion.id_persona
GROUP BY
	personas.id, personas.nombre,personas.apellidos
HAVING
	COUNT(persona_pertenece_faccion.id_faccion) > 1;

 * postgresql://postgres:***@localhost:5432/tarea1_bd
34 rows affected.


[(508, 'Shawnta', 'Schmidt'),
 (409, 'Rosa', 'Guffey'),
 (3, 'Samuel', 'Nason'),
 (984, 'Carolyn', 'Billings'),
 (639, 'Anthony', 'Schmitt'),
 (947, 'Tom', 'Williams'),
 (498, 'Twyla', 'Ona'),
 (838, 'Mary', 'Suggett'),
 (149, 'Yasmin', 'Kuster'),
 (717, 'Samuel', 'Williams'),
 (48, 'Jennifer', 'Allard'),
 (785, 'Jarrod', 'Donart'),
 (373, 'Paula', 'Tygart'),
 (289, 'Tara', 'Strassel'),
 (754, 'Teresa', 'Ridgway'),
 (65, 'Charles', 'Smith'),
 (225, 'Eduardo', 'Collar'),
 (207, 'Jocelyn', 'Tonge'),
 (912, 'John', 'Gutierrez'),
 (961, 'Robert', 'Davis'),
 (506, 'Truman', 'Liv'),
 (341, 'Maria', 'Simas'),
 (716, 'Jay', 'Gilbert'),
 (731, 'Eugene', 'Wilson'),
 (437, 'Clarisa', 'Norris'),
 (859, 'Claude', 'Wales'),
 (767, 'Robert', 'Appel'),
 (857, 'Jerry', 'Stern'),
 (268, 'Randall', 'Ellis'),
 (82, 'Charline', 'Nichols'),
 (216, 'Mary', 'Courtemanche'),
 (694, 'Marilyn', 'Warner'),
 (221, 'Lucille', 'Matos'),
 (649, 'Bonnie', 'Mclaren')]

### Consulta BONUS (6 pts.)

**Nombre y Apellidos de las personas casadas**(solo se muestra una única vez la pareja)

Ejemplo:
   
|Nombre persona 1 | Apellidos persona 1 | Nombre persona 2 | Apellidos persona 2 |
|-|-------|-|-|
| Felipe | Martinez | Raul | Cuello |
| Rodrigo | Munita | Juan | Mira |
| Nicole | Rojel | Benjamin | Campos |
| ... | ... | ... | ... |

**Solución:** _Escriba aquí la descripción de su solución._

In [22]:
%%sql

/* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%sql
SELECT
    * 
FROM 
    *

 * postgresql://postgres:***@localhost:5432/tarea1_bd
(psycopg2.errors.SyntaxError) error de sintaxis en o cerca de «%%»
LINE 3: %%sql
        ^

[SQL: /* Reemplace estas queries por las que resuelven las consultas pedidas. */

%%%%sql
SELECT
    * 
FROM 
    *]
(Background on this error at: https://sqlalche.me/e/20/f405)
